ssl/quic/quic_record_rx.c

```diff
@@ -0,0 +1,1164 @@
+
+    /* Feed AAD data. */
+    if (EVP_CipherUpdate(cctx, NULL, &l, aad, aad_len) != 1)
+        return 0;
+
+    /* Feed encrypted packet body. */
+    if (EVP_CipherUpdate(cctx, dst, &l, src, src_len - el->tag_len) != 1)
+        return 0;
+
+    /* Ensure authentication succeeded. */
+    if (EVP_CipherFinal_ex(cctx, NULL, &l2) != 1) {
+        /* Authentication failed, increment failed auth counter. */
+        ++qrx->forged_pkt_count;
+        return 0;
+    }
+
+    *dec_len = l;
+    return 1;
+}
+
+static ossl_inline void ignore_res(int x)
+{
+    /* No-op. */
+}
+
+static void qrx_key_update_initiated(OSSL_QRX *qrx)
+{
+    if (!ossl_qrl_enc_level_set_key_update(&qrx->el_set, QUIC_ENC_LEVEL_1RTT))
+        return;
+
+    if (qrx->key_update_cb != NULL)
+        qrx->key_update_cb(qrx->key_update_cb_arg);
+}
+
+/* Process a single packet in a datagram. */
+static int qrx_process_pkt(OSSL_QRX *qrx, QUIC_URXE *urxe,
+                           PACKET *pkt, size_t pkt_idx,
+                           RXE **first_rxe,
+                           size_t datagram_len)
+{
+    RXE *rxe;
+    const unsigned char *eop = NULL;
+    size_t i, aad_len = 0, dec_len = 0;
+    PACKET orig_pkt = *pkt;
+    const unsigned char *sop = PACKET_data(pkt);
+    unsigned char *dst;
+    char need_second_decode = 0, already_processed = 0;
+    QUIC_PKT_HDR_PTRS ptrs;
+    uint32_t pn_space, enc_level;
+    OSSL_QRL_ENC_LEVEL *el = NULL;
+
+    /*
+     * Get a free RXE. If we need to allocate a new one, use the packet length
+     * as a good ballpark figure.
+     */
+    rxe = qrx_ensure_free_rxe(qrx, PACKET_remaining(pkt));
+    if (rxe == NULL)
+        return 0;
+
+    /* Have we already processed this packet? */
+    if (pkt_is_marked(&urxe->processed, pkt_idx))
+        already_processed = 1;
+
+    /*
+     * Decode the header into the RXE structure. We first decrypt and read the
+     * unprotected part of the packet header (unless we already removed header
+     * protection, in which case we decode all of it).
+     */
+    need_second_decode = !pkt_is_marked(&urxe->hpr_removed, pkt_idx);
+    if (!ossl_quic_wire_decode_pkt_hdr(pkt,
+                                       qrx->short_conn_id_len,
+                                       need_second_decode, &rxe->hdr, &ptrs))
+        goto malformed;
+
+    /*
+     * Our successful decode above included an intelligible length and the
+     * PACKET is now pointing to the end of the QUIC packet.
+     */
+    eop = PACKET_data(pkt);
+
+    /*
+     * Make a note of the first RXE so we can later ensure the destination
+     * connection IDs of all packets in a datagram mater.
+     */
+    if (pkt_idx == 0)
+        *first_rxe = rxe;
+
+    /*
+     * Early header validation. Since we now know the packet length, we can also
+     * now skip over it if we already processed it.
+     */
+    if (already_processed
+        || !qrx_validate_hdr_early(qrx, rxe, pkt_idx == 0 ? NULL : *first_rxe))
```

Comment:
Not quite. We skip packets we already processed, but we do need to parse the headers minimally again to get the length so we can skip over it.

Reply:
Then I still do not understand the condition as currently if already processed the packet then we `goto malformed`?

Reply:
If we have already processed the packet we set the flag `already_processed` so that we skip processing of the packet *after* we have discerned its length, which is the minimum we need to be able to skip over it.

We set `eop` to point to the end of the current packet before the above line. The label `malformed:` handles things accordingly. If we have eop as non-NULL we skip to the next packet as we were able to discern the packet length. Otherwise we have no idea if another packet follows or not and just abandon processing.

I guess the label `malformed` could be considered misleading? It's really `malformed_or_already_processed`.

Reply:
Would it make sense to instead do here:
```
if (already_processed) {
    ignore_res(PACKET_forward(pkt, eop - PACKET_data(pkt)));
    return 0;
}
```
Or at least add a comment like `/* We treat already_processed packets in dgram like malformed packets, i.e., ignoring them. */`

Reply:
Added comment. I don't think it's worth optimising out the call to pkt_mark since it just sets a bit.

Reply:
I do not see the comment being pushed.

Reply:
Hmm, now I see it, I probably did not reload the GH view.